extensions/wikidata/src/org/openrefine/wikidata/qa/scrutinizers/EditScrutinizer.java

```diff
@@ -57,6 +54,19 @@ public void setStore(QAWarningStore store) {
     public void setFetcher(ConstraintFetcher fetcher) {
         _fetcher = fetcher;
     }
+
+    public void setConstraints(Constraints constraints) {
+        this.constraints = constraints;
+    }
+
+    /**
+     * Prepare the dependencies(i.e. constraint-related pids and qids) needed by the scrutinizer.
+     *
+     * Called before {@link EditScrutinizer#batchIsBeginning()}.
+     *
+     * @return false if any necessary dependency is missing, true otherwise.
+     */
+    public abstract boolean prepareDependencies();
```

Comment:
> I'm not sure I fully understand the lifecycle here, but could this be rolled into `batchIsBeginning()`? 

I don't think so. `batchIsBeginning()`, `scrutinize()` and `batchIsFinished()` are all parts of the scrutinization actually, see [NoEditsMadeScrutinizer](https://github.com/OpenRefine/OpenRefine/blob/master/extensions/wikidata/src/org/openrefine/wikidata/qa/scrutinizers/NoEditsMadeScrutinizer.java). My thought is that if `prepareDependencies` fails, the scrutinizer should not be registered at all, see my code from [EditInspector](https://github.com/afkbrb/OpenRefine/blob/wikibase-support/extensions/wikidata/src/org/openrefine/wikidata/qa/EditInspector.java) below:

```java
    public void register(EditScrutinizer scrutinizer) {
        scrutinizer.setStore(warningStore);
        scrutinizer.setFetcher(fetcher);
        scrutinizer.setConstraints(constraints);
        if (scrutinizer.prepareDependencies()) {
            String key = scrutinizer.getClass().getName();
            scrutinizers.put(key, scrutinizer);
        }
    }
```

> is it something that we want subclasses to explicitly think about and handle?

Yes. 

> Under what circumstances `prepareDependencies` fail? What should the user do about it?

`prepareDependencies ` fails if any dependency needed is missing (and this is expected). In this case, the scrutinizer is skipped since it can't work properly without dependencies. The user is not involved here.